### PR TITLE
Added osm-edge lightweight SMI compliant service mesh for edge

### PR DESCRIPTION
**Pull request by naqvis**
### [osm-edge](https://github.com/flomesh-io/osm-edge/) - a lightweight service mesh for the edge-computing.

This PR adds [osm-edge](https://github.com/flomesh-io/osm-edge/) addon integration to microk8s

*Also verify you have:*
* [X] Read the [contributions](https://github.com/ubuntu/microk8s/blob/master/CONTRIBUTING.md) page.
* [X] Submitted the [CLA form](https://ubuntu.com/legal/contributors/agreement), if you are a first time contributor.

### CODE REVIEW

**Comment by naqvis** [comment]
Not seeing any activity, please advise if i'm missing some procedure to get it moved forward?

**Comment by ktsakalozos** [comment]
Thank you for this PR @naqvis. We will take look at it shortly.

**Comment by ktsakalozos** [comment]
In the meantime could you address the lint/style issues reported in the tests?

**Comment by naqvis** [comment]
> In the meantime could you address the lint/style issues reported in the tests?

Thanks @ktsakalozos . I've fixed linting issue

**Comment by ktsakalozos** [comment]
@naqvis I am not able to get the tests passing in the respective github action but on an AWS this works fine. Are there any special requirements for the workload that does not allow it to work properly on the VM provided by the github runners?

**Comment by naqvis** [comment]
> @naqvis I am not able to get the tests passing in the respective github action but on an AWS this works fine. Are there any special requirements for the workload that does not allow it to work properly on the VM provided by the github runners?

Might be due to default installation installs some extra pods and that causes github action VM to overload. I'll be pushing changed script, which just install the service mesh only and see if github allocated vm is able to handle that load.

**Comment by naqvis** [comment]
@ktsakalozos  I looked at the action logs and notice Istio/Knative etc tests are being skipped. Believe they are being skipped due to github vm unable to handle the load. So question will be, can we also have this test skipped in github action?

**Comment by ktsakalozos** [comment]
@naqvis sure, we could disable the tests on GH actions but I still need to understand why we get these failures. I would not want to have this PR accepted and then reverted because it fails at a later stage of our CI pipeline.

**Comment by naqvis** [comment]
> @naqvis sure, we could disable the tests on GH actions but I still need to understand why we get these failures. I would not want to have this PR accepted and then reverted because it fails at a later stage of our CI pipeline.

Test case as you have commented before that works on normal VM or cloud instance. Its just Github action allocated VM is unable to handle the load, or else I would have expected all service mesh related testcases to be turned on as well. To confirm the reasoning, i would appreciate if you can try to turn on any or all of service mesh addon related tests and see if they passes the test cases on github action runner.

**Comment by naqvis** [comment]
@ktsakalozos looking at the github action runner logs, seems script is experiencing access privileges https://github.com/canonical/microk8s-community-addons/runs/7733177364?check_suite_focus=true#step:5:151

image


`enable` script is invoking utility method `run_with_sudo`, does that require some other privileges? or this method shouldn't be used?

File: addons/osm-edge/disable [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
```
Comment (ktsakalozos) [inline addons/osm-edge/disable:1]: When the addon is disabled the `pod/osm-controller` is not removed so its status is always set to enabled.

File: addons/osm-edge/enable [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+ARCH=$(arch)
+
+# check if osm-edge cli is already in the system.  Download if it doesn't exist.
+if [ ! -f "${SNAP_DATA}/bin/osm" ]; then
+  OSM_EDGE_VERSION="${OSM_EDGE_VERSION:-v1.1.0}"
+  echo "Fetching osm-edge version $OSM_EDGE_VERSION."
+
+  run_with_sudo mkdir -p "${SNAP_DATA}/tmp/osm-edge"
+  (cd "${SNAP_DATA}/tmp/osm-edge"
```
Comment (ktsakalozos) [inline addons/osm-edge/enable:17]: We are missing the closing parenthesis.

File: addons/osm-edge/disable (hunk 1 above)
Comment (naqvis) [reply addons/osm-edge/disable:1]: this shouldn't be the case, as `disable` script invoks `osm uninstall mesh` which will remove controller and other related pods. I couldn't reproduce this behavior on local boxes, can you please advise the steps which you followed to see this behavior?

File: addons/osm-edge/enable (hunk 2 above)
Comment (naqvis) [reply addons/osm-edge/enable:17]: Thanks for the catch.  I've fixed missing parenthesis issue.

File: addons/osm-edge/disable (hunk 1 above)
Comment (ktsakalozos) [reply addons/osm-edge/disable:1]: Here is what I get when I try to use osm:
```
ubuntu@ip-172-31-13-141:~$ /var/snap/microk8s/current/bin/osm uninstall mesh

List of meshes present in the cluster:
Unable to list meshes within the cluster.
Error: Could not list deployments Get "http://localhost:8080/apis/apps/v1/deployments?labelSelector=app%3Dosm-controller": dial tcp 127.0.0.1:8080: connect: connection refused
```

Is it possible the osm is trying to find the kubeconfig file in a defualt location?

Comment (ktsakalozos) [reply addons/osm-edge/disable:1]: It seems so:
```
ubuntu@ip-172-31-13-141:~$ mkdir ~/.kube
ubuntu@ip-172-31-13-141:~$ sudo microk8s.config > ~/.kube/config
ubuntu@ip-172-31-13-141:~$ /var/snap/microk8s/current/bin/osm uninstall mesh

List of meshes present in the cluster:

MESH NAME   MESH NAMESPACE   VERSION   ADDED NAMESPACES
osm         osm-system       v1.1.0    

MESH NAME   MESH NAMESPACE   SMI SUPPORTED
osm         osm-system       HTTPRouteGroup:v1alpha4,TCPRoute:v1alpha4,TrafficSplit:v1alpha2,TrafficTarget:v1alpha3

To list the OSM controller pods for a mesh, please run the following command passing in the mesh's namespace
        kubectl get pods -n <osm-mesh-namespace> -l app=osm-controller

Uninstall OSM [mesh name: osm] in namespace [osm-system] and/or OSM resources ? [y/n]: ^C
```

Comment (ktsakalozos) [reply addons/osm-edge/disable:1]: Here is something that worked:
```
sudo KUBECONFIG=/var/snap/microk8s/current/credentials/client.config  /var/snap/microk8s/current/bin/osm uninstall mesh
```

Comment (naqvis) [reply addons/osm-edge/disable:1]: Yeah, just figured that `disable` script was missing the location of `KUBECONFIG`. I've updated `disable` script and added `KUBECONFIG`.

**Comment by ktsakalozos** [comment]
Thank you for this PR @naqvis. This addon will be available on the upcoming release v1.25.

One last thing we would like your help with is the docs page that will be linked from [1]. I created a placeholder page for now [2]. Please let us know how would you like this page to look like.

Thanks you!

[1] https://microk8s.io/docs/addons
[2] https://discuss.kubernetes.io/t/addon-osm-edge/20969

**Comment by naqvis** [comment]
> Thank you for this PR @naqvis. This addon will be available on the upcoming release v1.25.
> 
> One last thing we would like your help with is the docs page that will be linked from [1]. I created a placeholder page for now [2]. Please let us know how would you like this page to look like.
> 
> Thanks you!
> 
> [1] https://microk8s.io/docs/addons [2] https://discuss.kubernetes.io/t/addon-osm-edge/20969

Thank you @ktsakalozos for your help and support.

For placeholder part, is there a repo against which I need to submit PR? or I can provide a one liner description to be included in [1]?

For discussion forum [2], your provided description is good enough. Please advise if we are required to follow some template? or specific procedure to have it updated.

[1] https://microk8s.io/docs/addons
[2] https://discuss.kubernetes.io/t/addon-osm-edge/20969

**Pull request merged**